src/Raven.Server/ServerWide/Maintenance/ClusterObserver.cs

```diff
@@ -179,191 +175,237 @@ class ClusterObserver : IDisposable
+            if (status == ConflictStatus.AlreadyMerged)
+            {
+                if (previous.TryGetValue(promotable, out var promotablePrevClusterStats) == false ||
+                    promotablePrevClusterStats.LastReport.TryGetValue(dbName, out var promotablePrevDbStats) == false)
+                    return false;
+
+                var indexesCatchedUp = CheckIndexProgress(promotablePrevDbStats.LastEtag, promotablePrevDbStats.LastIndexStats, promotableDbStats.LastIndexStats);
+                if (indexesCatchedUp)
+                {
+                    topology.Promotables.Remove(promotable);
+                    topology.Members.Add(promotable);
+                    if (_logger.IsOperationsEnabled)
+                    {
+                        _logger.Operations($"We promote the database {dbName} on {promotable} to be a full member");
+                    }
+                    topology.PromotablesStatus.Remove(promotable);
+                    topology.DemotionReasons.Remove(promotable);
+
+                    return true;
+                }
+                if (_logger.IsInfoEnabled)
+                {
+                    _logger.Info($"The database {dbName} on {promotable} not ready to be promoted, because the indexes are not up-to-date.\n");
+                }
 
+                topology.PromotablesStatus[promotable] = "node is not ready to be promoted, because the indexes are not up-to-date";
+            }
+            else
+            {
+                if (_logger.IsInfoEnabled)
+                {
+                    _logger.Info($"The database {dbName} on {promotable} not ready to be promoted, because the change vectors are {status}.\n" +
+                                 $"mentor's change vector : {mentorPrevDbStats.LastChangeVector}, node's change vector : {promotableDbStats.LastChangeVector}");
                 }
+                topology.PromotablesStatus[promotable] = $"node is not ready to be promoted, because the change vectors are {status}.\n" +
+                                                         $"mentor's change vector : {mentorPrevDbStats.LastChangeVector}, " +
+                                                         $"node's change vector : {promotableDbStats.LastChangeVector}";
             }
             return false;
         }
 
-        private const DatabaseStatus BadFlags = Faulted | Shutdown;
-        private bool ShouldReassign(
+        private void RemoveOtherNodesIfNeeded(DatabaseTopology topology)
+        {
+            if (topology.Members.Count == topology.ReplicationFactor)
+            {
+                if (topology.Promotables.Count + topology.Rehab.Count > 0)
+                {
+                    if (_logger.IsOperationsEnabled)
+                    {
+                        _logger.Operations("We reached the replication factor, so we remove all other broken/promotable nodes.");
+                    }
+                    topology.Promotables.Clear();
```

Comment:
This is wrong, we need to add them to `DeletionInProgress` so they'll delete the data properly. See RavenDB-7913